cmd/otel-allocator/main.go

```diff
@@ -16,306 +16,205 @@ package main
-	if err != nil {
-		setupLog.Error(err, "Unable to load configuration")
+	cfg, configLoadErr := config.Load(*cliConf.ConfigFilePath)
+	if configLoadErr != nil {
+		setupLog.Error(configLoadErr, "Unable to load configuration")
 	}
 
 	cliConf.RootLogger.Info("Starting the Target Allocator")
-
 	ctx := context.Background()
-
 	log := ctrl.Log.WithName("allocator")
 
-	// allocatorPrehook will be nil if filterStrategy is not set or
-	// unrecognized. No filtering will be used in this case.
-	allocatorPrehook := prehook.New(cfg.GetTargetsFilterStrategy(), log)
-
-	allocator, err := allocation.New(cfg.GetAllocationStrategy(), log, allocation.WithFilter(allocatorPrehook))
+	events = make(chan allocatorWatcher.Event)
+	errors = make(chan error)
+	allocatorPrehook = prehook.New(cfg.GetTargetsFilterStrategy(), log)
+	allocator, err = allocation.New(cfg.GetAllocationStrategy(), log, allocation.WithFilter(allocatorPrehook))
 	if err != nil {
 		setupLog.Error(err, "Unable to initialize allocation strategy")
 		os.Exit(1)
 	}
-
-	watcher, err := allocatorWatcher.NewWatcher(setupLog, cfg, cliConf, allocator)
-	if err != nil {
-		setupLog.Error(err, "Can't start the watchers")
+	discoveryCtx, discoveryCancel := context.WithCancel(ctx)
+	discoveryManager = discovery.NewManager(discoveryCtx, gokitlog.NewNopLogger())
+	targetDiscoverer = target.NewDiscoverer(log, discoveryManager, allocatorPrehook)
+	collectorWatcher, collectorWatcherErr := collector.NewClient(log, cliConf.ClusterConfig)
+	if collectorWatcherErr != nil {
+		setupLog.Error(collectorWatcherErr, "Unable to initialize collector watcher")
 		os.Exit(1)
 	}
-	defer func() {
-		watcherErr := watcher.Close()
-		if watcherErr != nil {
-			log.Error(watcherErr, "failed to close watcher")
-		}
-	}()
-
-	// creates a new discovery manager
-	discoveryManager := lbdiscovery.NewManager(ctx, log, gokitlog.NewNopLogger(), allocatorPrehook)
-	defer discoveryManager.Close()
-
-	discoveryManager.Watch(allocator.SetTargets)
-
-	k8sclient, err := configureFileDiscovery(context.Background(), log, allocator, discoveryManager, cliConf)
+	fileWatcher, err = allocatorWatcher.NewFileWatcher(setupLog.WithName("file-watcher"), cliConf)
 	if err != nil {
-		setupLog.Error(err, "Can't start the k8s client")
+		setupLog.Error(err, "Can't start the file watcher")
 		os.Exit(1)
 	}
-
-	srv := newServer(log, allocator, discoveryManager, k8sclient, cliConf.ListenAddr)
-
-	interrupts := make(chan os.Signal, 1)
-	signal.Notify(interrupts, os.Interrupt, syscall.SIGTERM)
-
-	go func() {
-		if err := srv.Start(); !errors.Is(err, http.ErrServerClosed) {
-			setupLog.Error(err, "Can't start the server")
+	if *cliConf.PromCRWatcherConf.Enabled {
+		promWatcher, err = allocatorWatcher.NewPrometheusCRWatcher(cfg, cliConf)
+		if err != nil {
+			setupLog.Error(err, "Can't start the prometheus watcher")
+			os.Exit(1)
 		}
-	}()
-
-	for {
-		select {
-		case <-interrupts:
+		promWatcherErr := promWatcher.Start(events, errors)
```

Comment:
It's because of how Prometheus CRD watching is setup – each informer (podmonitor, servicemonitor) is already async so we can't make them match easily unless we just wanna throw a while loop around the promwatcher method.